test/test.lua

```diff
@@ -1634,6 +1634,93 @@ function test.multi_gpu_random()
    cutorch.setDevice(1) -- reset device
 end
 
+function test.multinomial_with_replacement()
+   for tries = 1, 10 do
+      local n_row = torch.random(10)
+      local n_col = 1 + torch.random(1000)
+
+      local prob_dist = torch.CudaTensor(n_row, n_col):uniform()
+      prob_dist:select(2, n_col):fill(0) --index n_col shouldn't be sampled
+      local n_sample = torch.random(n_col - 1)
+      local sample_indices = torch.multinomial(prob_dist, n_sample, true)
+      tester:assert(prob_dist:dim() == 2, "wrong number of prob_dist dimensions")
+      tester:assert(sample_indices:size(2) == n_sample, "wrong number of samples")
+
+      for i = 1, n_row do
+         for j = 1, n_sample do
+            local val = sample_indices[{i,j}]
+            tester:assert(val == math.floor(val) and val >= 1 and val < n_col,
+                          "sampled an invalid index: " .. val)
+         end
+      end
+   end
+end
+
+function test.multinomial_without_replacement()
+   for tries = 1, 10 do
+      local n_row = torch.random(1000)
+      -- choose a small number of columns to test that the 0 col is never chosen
+      local n_col = 1 + torch.random(10)
+
+      local prob_dist = torch.CudaTensor(n_row, n_col):uniform()
+      prob_dist:select(2, n_col):fill(0) --index n_col shouldn't be sampled
+      local n_sample = torch.random(n_col - 1)
+      local sample_indices = torch.multinomial(prob_dist, n_sample, false)
+      tester:assert(prob_dist:dim() == 2, "wrong number of prob_dist dimensions")
+      tester:assert(sample_indices:size(2) == n_sample, "wrong number of samples")
+
+      for i = 1, n_row do
+         local row_samples = {}
+         for j = 1, n_sample do
+            local sample_idx = sample_indices[{i,j}]
```

Comment:
This will be rather slow. Each index triggers a memcpy from the GPU which has high latency. It should be faster to copy the entire tensor to the CPU and then do the checks on that.